### PR TITLE
Development

### PR DESCRIPTION
**Pull request by diaabadr**

### CODE REVIEW

File: iniparser.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+type Section map[string]string
+
+type Ini map[string]Section
+
+type Parser struct {
+	sections Ini
+}
+
+func NewINI() Ini {
```
Comment (xmonader) [inline iniparser.go]: This is wrong, It should be NewINIParser that returns Parser struct

File: README.md [hunk 2]
```diff
@@ -1 +1,78 @@
-# iniparser-Diaa
\ No newline at end of file
+# INI Parser for Go
+
+This package provides a parser for INI files in Go.
+
+# Parsing INI Data
+To parse INI data, create a new Parser object using the NewParser() function:
+```
+parser := iniparser.NewParser()
```
Comment (rawdaGastan) [inline README.md:12]: make sure to add how to import the package `in case someone is copying the code`

File: README.md [hunk 3]
```diff
@@ -1 +1,78 @@
-# iniparser-Diaa
\ No newline at end of file
+# INI Parser for Go
```
Comment (rawdaGastan) [inline README.md:1]: That's great. 

we need to add `how to test (test section)`

File: README.md [hunk 4]
```diff
@@ -1 +1,78 @@
...
+
+This package provides a parser for INI files in Go.
+
+# Parsing INI Data
+To parse INI data, create a new Parser object using the NewParser() function:
```
Comment (rawdaGastan) [inline README.md:7]: adding an INI example would be great,
for example:

```json
[owner]
name = John
organization = threefold

[database]
version = 12.6
server = 192.0.2.62
port = 143
password = 123456
protected = true
```

File: .github/workflows/go-test.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+name: Go Test
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+
+jobs:
+  test:
```
Comment (rawdaGastan) [inline .github/workflows/go-test.yaml:10]: It's ok. but I think it would be better to merge test and lint jobs to prevent repeated steps (less time)

we can also add staticcheck `dominikh/staticcheck-action` and gofmt `Jerome1337/gofmt-action`

`golangci-lint` is also good

File: INI-files/config.ini [hunk 6]
```diff
@@ -0,0 +1,9 @@
+[new[section]]
```
Comment (rawdaGastan) [inline INI-files/config.ini]: I think It would be better if this can be a great example in the readme and we can remove this directory.

File: go.mod [hunk 7]
```diff
@@ -0,0 +1,5 @@
+module iniparser
+
+go 1.20
+
+require github.com/pkg/errors v0.9.1
```
Comment (rawdaGastan) [inline go.mod]: files should have an empty line as its last line.

File: iniparser.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// The error messages that can be returned by the INI parser.
+var (
+	ErrInvalidFormat    = errors.New("invalid format ")
```
Comment (rawdaGastan) [inline iniparser.go]: make sure to add a comment before each exported value.

File: iniparser.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+			key, value = strings.TrimSpace(key), strings.TrimSpace(value)
+
+			if len(key) == 0 {
+				return fmt.Errorf("%w invalid key at line %d", ErrInvalidFormat, idx)
+			}
+
+			// Add key-value pair to current section
+			p.sections[currentSection][key] = value
+		} else {
+			return fmt.Errorf("%w invalid format at line %d", ErrInvalidFormat, idx)
```
Comment (rawdaGastan) [inline iniparser.go]: I think `invalid format` will be repeated because `ErrInvalidFormat` is already mentioning it.

File: iniparser.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+		// Check for section header
+		if strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") {
+
+			// Extract section name
+			sectionName := line[1 : len(line)-1]
+			sectionName = strings.TrimSpace(sectionName)
+
+			if len(sectionName) == 0 {
+				return fmt.Errorf("%w invalid section at line %d", ErrInvalidFormat, idx)
```
Comment (rawdaGastan) [inline iniparser.go]: It would be better if it's `%w: ....`

File: iniparser.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+		// Check for key-value pairs
+		if strings.Contains(line, "=") && currentSection != "" {
+
+			// Parse key and value
+			keyAndValue := strings.Split(line, "=")
+			key, value := keyAndValue[0], keyAndValue[1]
+			key, value = strings.TrimSpace(key), strings.TrimSpace(value)
+
+			if len(key) == 0 {
+				return fmt.Errorf("%w invalid key at line %d", ErrInvalidFormat, idx)
```
Comment (rawdaGastan) [inline iniparser.go]: It would be better if it's `%w: ....`

File: iniparser.go [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+		} else {
+			return fmt.Errorf("%w invalid format at line %d", ErrInvalidFormat, idx)
+		}
+	}
+	return scanner.Err()
+}
+
+// GetSectionNames returns the names of all sections in the INI data.
+func (p *Parser) GetSectionNames() []string {
+	sections := make([]string, 0, len(p.sections))
```
Comment (rawdaGastan) [inline iniparser.go:130]: FYI: `var sections []string`, `sections := []string{}` are also good

File: iniparser.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+			configText += fmt.Sprintf("%s=%s\n", key, value)
+		}
+	}
+
+	return configText
+}
+
+// SaveToFile saves the INI data to a file.
+func (p *Parser) SaveToFile(path string) error {
+
```
Comment (rawdaGastan) [inline iniparser.go:192]: why not checking `ini` extension?

File: iniparser_test.go [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+spaces in keys=allowed
+spaces in values=allowed as well
+[Complex Values]
+spaces around the delimiter=obviously
+key=
+[new[section]]
+new key = new value
+`
+
+var iniInvalidFormats = []string{
```
Comment (rawdaGastan) [inline iniparser_test.go:27]: we can add a test for an empty section name `[]`

File: iniparser_test.go [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+	})
+}
+
+func assertContainsSubString(input, out, target string) bool {
+
+	if !strings.Contains(input, target) || !strings.Contains(out, target) {
+		return false
+	}
+	return true
+}
```
Comment (rawdaGastan) [inline iniparser_test.go]: add an empty last line :D

hint: format extensions (for go, md files, json, yml) in vscode will help you

File: iniparser_test.go [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+
+func TestLoadFromString(t *testing.T) {
+	p := NewParser()
+
+	t.Parallel()
+
+	t.Run("Valid INI Syntax", func(t *testing.T) {
+		err := p.LoadFromString(iniValidFormat)
+
+		if !errors.Is(err, nil) {
```
Comment (rawdaGastan) [inline iniparser_test.go]: FYI: try `github.com/stretchr/testify/assert` in your upcoming tests. It will make checks easier.

an example: we can replace those line with just `assert.NoError(t, err)` only
`assert.Equals(t, w1, w2)`
`assert.Contains(t, w, sub)`
.....

File: iniparser_test.go [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+
+	t.Parallel()
+
+	t.Run("Valid File Path", func(t *testing.T) {
+
+		file, err := os.CreateTemp("", "config-*.ini")
+		if err != nil {
+			t.Errorf("Error creating temporary file: %q", err)
+		}
+		defer os.Remove(file.Name())
```
Comment (rawdaGastan) [inline iniparser_test.go]: we can use `tempDir`. It will remove the whole dir in the end of the test by itself

File: iniparser_test.go [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+	})
+}
+
+func assertStrings(t testing.TB, got, want string) {
+	if got != want {
+		t.Errorf("got %q want %q", got, want)
+	}
+}
+
+func TestSaveToFile(t *testing.T) {
```
Comment (rawdaGastan) [inline iniparser_test.go:212]: don't forget to add a test for the extension when you add it :)

**Review by rawdaGastan — CHANGES_REQUESTED** [review]
Great work. thank you!

I added some comments/improvements. kindly check them

File: go.mod [hunk 19]
```diff
@@ -0,0 +1,5 @@
+module iniparser
```
Comment (rawdaGastan) [inline go.mod]: It's better to name the module as `github.com/codescalersinternships/iniparser-Diaa`
so it will be related to the repo for importing purposes

File: iniparser.go [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+	scanner := bufio.NewScanner(reader)
+	idx := 0
+	for scanner.Scan() {
+		idx++
+		line := scanner.Text()
+
+		line = strings.TrimSpace(line)
+
+		// Skip empty lines and comments
+		if len(line) == 0 || string(line[0]) == ";" {
```
Comment (rawdaGastan) [inline iniparser.go]: comments can start with a `#`

File: iniparser.go [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+		return err
+	}
+	defer file.Close()
+
+	return p.LoadFromReader(bufio.NewReader(file))
+}
+
+// LoadFromReader loads INI data from an io.Reader object.
+func (p *Parser) LoadFromReader(reader io.Reader) error {
+	var currentSection string = ""
```
Comment (rawdaGastan) [inline iniparser.go:75]: make sure your data map is empty before you start loading

**Review by rawdaGastan — APPROVED** [review]
